### ISSUE TITLE
Crash after connect with Airfoil

### ISSUE THREAD

**Issue by pgergis** [philippe44/AirConnect#134]
Not expecting that you have Airfoil, but I think it might turn out to be a fairly common use case for this tool, hence opening this issue; happy to test / debug / help however I can if you're able to help out! (Thanks for your work btw!)

Running `./airupnp-osx-multi -l 1000:2000` successfully finds and works with my old Sonos speaker. Logs:

```
[17:17:29.171622] main:1224 Starting airupnp version: v0.2.2.4 (Dec 16 2018 @ 16:04:50)
[17:17:29.172081] main:1232 no config file, using defaults
[17:17:29.174143] Start:984 Binding to 192.168.1.174:49152
[17:17:30.968028] AddMRDevice:896 [0x10b0509c0]: adding renderer (Studio)
[17:17:31.001014] MasterHandler:604 [0x10b0509c0]: subscribe success
[17:17:31.130841] ProcessEvent:403 [0x10b0509c0]: UPnP Volume local change 68
[17:17:42.049516] rtsp_thread:323 got RTSP connection 11
[17:17:42.050348] handle_rtsp:371 [0x7f8ebad000f0]: challenge ig16+aoPrvO+iin8z1D32Q
[17:17:42.054956] handle_rtsp:364 [0x7f8ebad000f0]: received ANNOUNCE
[17:17:42.057443] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 2
```

When I launch [Airfoil](https://rogueamoeba.com/airfoil/) (a popular solution for grouping / streaming to multiple airplay 1 speakers), and try to stream to the Sonos speaker, I get the following logs before AirConnect crashes (apparently without cleanup; the Sonos speaker continues showing up as an Airplay device):

```
[17:17:42.058459] handle_rtsp:364 [0x7f8ebad000f0]: received SETUP
[17:17:42.058647] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Transport: RTP/AVP/UDP;unicast;mode=record;control_port=50410;timing_port=49547;server_port=60556
Session: DEADBEEF
Audio-Jack-Status: connected; type=analog
CSeq: 3


[17:17:42.059492] handle_rtsp:364 [0x7f8ebad000f0]: received RECORD
[17:17:42.059501] callback:273 [0x10b0509c0]: Stream
[17:17:42.059512] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Audio-Latency: 44100
Audio-Jack-Status: connected; type=analog
CSeq: 4


[17:17:42.060634] handle_rtsp:364 [0x7f8ebad000f0]: received SET_PARAMETER
[17:17:42.060643] handle_rtsp:519 [0x7f8ebad000f0]: SET PARAMETER volume -11.250000
[17:17:42.060647] CtrlSetVolume:253 [0x10b0509c0]: uPNP volume 62 (cookie 0x0)
[17:17:42.060720] callback:331 [0x10b0509c0]: Volume[0..100] 62
[17:17:42.060734] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 5


[17:17:42.066811] flac_init:174 [0x7f8ebc003e00]: Using FLAC-0 (0x7f8ebd300030)
[17:17:42.066984] buffer_put_packet:507 [0x7f8ebc003e00]: fill status [level:65535] [W:630 R:631]
[17:17:42.068651] handle_rtsp:364 [0x7f8ebad000f0]: received SET_PARAMETER
[17:17:42.068672] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 6


[17:17:42.069609] handle_rtsp:364 [0x7f8ebad000f0]: received SET_PARAMETER
[17:17:42.069649] handle_rtsp:536 [0x7f8ebad000f0]: received metadata
	artist: Via Airfoil on Pierre’s iMac Pro
	album:  
	title:  USB Audio CODEC 
[17:17:42.069663] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 7


[17:17:42.078461] rtp_thread_func:643 [0x7f8ebc003e00]: 1st sync packet received
[17:17:42.084638] handle_rtsp:364 [0x7f8ebad000f0]: received SET_PARAMETER
[17:17:42.084668] handle_rtsp:551 [0x7f8ebad000f0]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 8


Bus error: 10
```

I then have to relaunch and properly shutdown AirConnect to get the Sonos speaker to stop showing up as an Airplay speaker.

Any thoughts on why the crash happens / how I can help debug this? Apologies for opening this fairly novice / low-tech issue; like I said, happy to help however I can, but no worries if this is too niche to deal with right now.

Thanks!

**Comment by avggeek** [comment]
Hello - just wanted to add on that I too am seeing this problem with Airfoil. Here is my current setup:

- AirConnect environment:
```
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"

$ ./airupnp-x86-64 -h
v0.2.2.4 (Dec 16 2018 @ 16:04:42)
```

- AirFoil environment:
```
OS Version: Windows 10 Home Version 1803
AirFoil version: v5.6 
```

- Steps to Reproduce:

1. Run AirConnect on Debian host
2. Run AirFoil on Windows 10 PC
3. Verify that Sonos speaker appears in the list of available speakers.
3. Begin playback of a track in Spotify
4. In AirFoil, activate the Sonos speaker (Music note icon should turn blue)

- Expected Behaviour:
Music playback will start from Sonos speaker

- Actual Behaviour:
AirConnect crashes with a Segmentation Fault - console output below:
```
$ ./airupnp-x86-64
[17:03:51.824740] main:1224 Starting airupnp version: v0.2.2.4 (Dec 16 2018 @ 16:04:42)
[17:03:51.824782] main:1232 no config file, using defaults
[17:03:51.879064] Start:984 Binding to 192.168.50.25:49152
[17:03:52.177658] AddMRDevice:896 [0x6a5e80]: adding renderer (Portable)
[17:03:52.188109] MasterHandler:604 [0x6a5e80]: subscribe success
[17:03:52.211939] ProcessEvent:403 [0x6a5e80]: UPnP Volume local change 100
[17:05:44.385961] ProcessEvent:403 [0x6a5e80]: UPnP Volume local change 85
[17:05:44.651003] ProcessEvent:403 [0x6a5e80]: UPnP Volume local change 46
[17:05:44.976546] ProcessEvent:403 [0x6a5e80]: UPnP Volume local change 43
[17:05:45.223082] ProcessEvent:403 [0x6a5e80]: UPnP Volume local change 45
[17:05:45.508714] ProcessEvent:403 [0x6a5e80]: UPnP Volume local change 55
[17:06:46.120499] rtsp_thread:323 got RTSP connection 12
[17:06:46.128379] handle_rtsp:371 [0x7f44f4003500]: challenge Pra5kq/aMqQNdQVw+zhx1g
[17:06:46.134828] handle_rtsp:371 [0x7f44f4003500]: challenge 9uq0xuf4KV1/+6B9NnDsoA
[17:06:46.136695] handle_rtsp:371 [0x7f44f4003500]: challenge 9uq0xuf4KV1/+6B9NnDsoA
[17:06:46.139919] handle_rtsp:364 [0x7f44f4003500]: received ANNOUNCE
[17:06:46.141063] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 4


[17:06:46.145170] handle_rtsp:364 [0x7f44f4003500]: received SETUP
[17:06:46.145847] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Transport: RTP/AVP/UDP;unicast;mode=record;control_port=46425;timing_port=59600;server_port=40715
Session: DEADBEEF
Audio-Jack-Status: connected; type=analog
CSeq: 5


[17:06:46.148284] handle_rtsp:364 [0x7f44f4003500]: received RECORD
[17:06:46.148295] callback:273 [0x6a5e80]: Stream
[17:06:46.148310] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 6


[17:06:46.149640] handle_rtsp:364 [0x7f44f4003500]: received SET_PARAMETER
[17:06:46.149650] handle_rtsp:519 [0x7f44f4003500]: SET PARAMETER volume -16.320000
[17:06:46.149663] CtrlSetVolume:253 [0x6a5e80]: uPNP volume 45 (cookie (nil))
[17:06:46.149719] callback:331 [0x6a5e80]: Volume[0..100] 45
[17:06:46.149798] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 7


[17:06:48.568749] rtp_thread_func:643 [0x7f44e4002230]: 1st sync packet received
[17:06:48.569999] flac_init:174 [0x7f44e4002230]: Using FLAC-0 (0x7f44d00008e0)
[17:06:48.570038] buffer_put_packet:507 [0x7f44e4002230]: fill status [level:65535] [W:42 R:43]
[17:06:48.570358] handle_rtsp:364 [0x7f44f4003500]: received SET_PARAMETER
[17:06:48.570378] handle_rtsp:536 [0x7f44f4003500]: received metadata
        artist:
        album:
        title:
[17:06:48.570395] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 8


[17:06:48.573039] handle_rtsp:364 [0x7f44f4003500]: received SET_PARAMETER
[17:06:48.573055] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 9


[17:06:48.574191] handle_rtsp:364 [0x7f44f4003500]: received SET_PARAMETER
[17:06:48.574202] handle_rtsp:536 [0x7f44f4003500]: received metadata
        artist:
        album:
        title:
[17:06:48.574216] handle_rtsp:551 [0x7f44f4003500]: responding:
RTSP/1.0 200 OK
Audio-Jack-Status: connected; type=analog
CSeq: 10


[17:06:48.618272] AVTSetURI:91 [0x6a5e80]: uPNP setURI http://192.168.50.25:49677/stream.flac (cookie 0x1)
[17:06:48.618375] AVTPlay:151 [0x6a5e80]: uPNP play (cookie 0x2)
[17:06:48.618391] CtrlSetVolume:253 [0x6a5e80]: uPNP volume 45 (cookie 0x2)
[17:06:48.685564] http_thread_func:919 [0x7f44e4002230]: got HTTP connection 18 (silent frames 0)
[17:06:48.685882] handle_http:1112 [0x7f44e4002230]: received GET
[17:06:48.685899] handle_http:1139 [0x7f44e4002230]: responding:
HTTP/1.0 200 OK
Server: HairTunes
Content-Type: audio/flac
Connection: close


[17:06:48.685907] _buffer_get_frame:839 [0x7f44e4002230]: drain status [level:9] [W:52 R:43] [R:0 S:0]
FIXME: unhandled predicition type: 13
FIXME: unhandled predicition type: 5
FIXME: unhandled predicition type: 13
FIXME: unhandled predicition type: 3
Segmentation fault
```

**Comment by philippe44** [comment]
Should be fixed in 0.2.2.5

**Comment by avggeek** [comment]
Hi @philippe44 - I have good news and bad news.

The good news is that the crash on starting playback with AirFoil is definitely resolved on 0.2.2.5. However I am seeing some strange behaviour with the latency settings and playback especially from iTunes.

- AirConnect environment
```
$ ./airupnp-x86-64 -h
v0.2.2.5 (Jan 12 2019 @ 15:17:32)
```

- Steps to Reproduce stuttering in AirFoil playback:
1. Run AirConnect on Debian Host with following settings
`/home/superuser/bin/airupnp-x86-64 -l 1000:2000`
 2. Run AirFoil on Windows 10 PC
3. Verify that Sonos speaker appears in the list of available speakers.
 4. Begin playback of a track in Spotify
 5. In AirFoil, activate the Sonos speaker (Music note icon should turn blue)

 - Expected Behaviour:
Music playback will start from Sonos speaker

- Actual Behaviour:
Playback either does not commence or is extremely choppy.
_Workaround_: Launch AirConnect with changes to _RTSP_ latency:
`/home/superuser/bin/airupnp-x86-64 -l 8000:2000`

- Steps to Reproduce stuttering in iTunes playback:
1. Run AirConnect on Debian Host with following settings
`/home/superuser/bin/airupnp-x86-64 -l 8000:2000`
 2. Run iTunes on Windows 10 PC
3. Verify that Sonos speaker appears in the list of available speakers.
 4. Select Sonos speaker from output list
 5. In iTunes begin playback

 - Expected Behaviour:
Music playback will start from Sonos speaker

- Actual Behaviour:
Playback is extremely choppy.
Changes to _RTSP_ latency or _HTTP_ latency have no effect:
`/home/superuser/bin/airupnp-x86-64 -l 8000:8000`

**Comment by philippe44** [comment]
oops ... a correction that I did on a common piece of code to handle the airplay of HTC connect disabled the use of silent frames (http_latency). Corrected in 0.2.2.6

**Comment by avggeek** [comment]
> 
> 
> oops ... a correction that I did on a common piece of code to handle the airplay of HTC connect disabled the use of silent frames (http_latency). Corrected in 0.2.2.6

Hi @philippe44 I'm happy to report that with 0.2.2.6 I'm able to stream audio to AirFoil without crashes and iTunes as well. In both cases using the default `./airupnp-x86-64 -l 1000:2000` gives me skip-free audio.

Thank you very much for releasing this software and for the super-quick fixes!

**Comment by pgergis** [comment]
+1, thanks so much for the quick fixes!
